### PR TITLE
Loopy sprint

### PR DESCRIPTION
**Pull request by ReubenHill**
Linked to https://github.com/firedrakeproject/firedrake/pull/2038 and https://github.com/firedrakeproject/tsfc/pull/244 . Requires loopy branch https://github.com/firedrakeproject/loopy/tree/kernel_callables_v3-edit2

### CODE REVIEW

**Comment by sv2518** [comment]
This is ready to be reviewed, @ReubenHill could you please convert into an actual (not draft) PR?

**Comment by ReubenHill** [comment]
> This is ready to be reviewed, @ReubenHill could you please convert into an actual (not draft) PR?

done

File: pyop2/base.py [hunk 1]
```diff
@@ -1561,7 +1561,7 @@ def zero(self, subset=None):
                 i = p.Variable("i")
                 insn = loopy.Assignment(x.index(i), 0, within_inames=frozenset(["i"]))
                 data = loopy.GlobalArg("dat", dtype=self.dtype, shape=(self.cdim,))
-                knl = loopy.make_function([domain], [insn], [data], name="zero")
+                knl = loopy.make_function([domain], [insn], [data], name="zero", target=loopy.CTarget(), lang_version=(2018, 2))
```
Comment (sv2518) [inline pyop2/base.py:1564]: Maybe we can set the lang version globally

File: pyop2/codegen/rep2loopy.py [hunk 2]
```diff
@@ -579,27 +535,34 @@ def generate(builder, wrapper_name=None):
     preamble = "\n".join(sorted(headers))
 
     from coffee.base import Node
+    from loopy.kernel.function_interface import CallableKernel
 
-    if isinstance(kernel._code, loopy.LoopKernel):
-        from loopy.transform.callable import _match_caller_callee_argument_dimension_
+    if isinstance(kernel._code, loopy.TranslationUnit):
+        from pyop2.codegen.loopycompat import _match_caller_callee_argument_dimension_
```
Comment (dham) [inline pyop2/codegen/rep2loopy.py:541]: Move this to the top of the file.

File: requirements-git.txt [hunk 3]
```diff
@@ -1,4 +1,2 @@
-git+https://github.com/firedrakeproject/petsc.git@firedrake#egg=petsc
---no-deps git+https://github.com/firedrakeproject/petsc4py.git@firedrake#egg=petsc4py
 git+https://github.com/coneoproject/COFFEE.git#egg=coffee
-git+https://github.com/firedrakeproject/loopy.git@firedrake#egg=loopy
+git+https://github.com/firedrakeproject/loopy.git@kernel_callables_v3-edit2#egg=loopy
```
Comment (dham) [inline requirements-git.txt:2]: Fix this.

**Review by dham — CHANGES_REQUESTED** [review]

**Pull request merged**